docs/functions.md

```diff
@@ -0,0 +1,209 @@
+                    | Results     +<--+ | Action | |
+                    +-------------+   | +--------+ |
+                                      +------------+
+```
+
+A function always:
+
+  1. Starts with a [**reader**](readers.md)
+  2. Operates on zero or more [**records**](overview.md#record)
+  3. Consists of zero or more [**operations**](operations.md) (a.k.a steps)
+  4. Ends with an [**action**](overview.md#action)
+  5. Returns zero or more [**results**](overview.md#result)
+  6. May generate zero or more errors
+
+## Execution
+A function is executed by the RedisGears engine in one of two ways:
+
+  1. **Batch**: execution is immediate and on existing data
+  2. **Event**: execution is triggered by new events and on their data
+
+The function's mode of execution is determined by its action. There are two types of actions:
+
+  1. [**Run**](#run): runs the function in batch
+  2. [**Register**](#register): registers the function to be triggered by events
+
+When executed, whether as batch or event, the function's context is managed by the engine. Besides the function's logic, the context also includes its breakdown to internal execution steps, state, status, statistics, results and any errors encountered among other things.
+
+!!! abstract "Related commands"
+    The following RedisGears commands are related to executing functions:
+
+    * [`RG.PYEXECUTE`](commands.md#rgpyexecute)
+    * [`RG.ABORTEXECUTION`](commands.md#rgabortexecution)
+    * [`RG.DUMPEXECUTIONS`](commands.md#rgdumpexecutions)
+    * [`RG.GETEXECUTION`](commands.md#rggetexecution)
+
+## Execution ID
+The execution of every function is internally assigned with a unique value called **Execution ID**.
+
+The ID is string value made of two parts that are delimited by a hyphen ('-'), as follows:
+
+  1. **Shard ID**: the 40-bytes-long identifier of a [shard](overview.md#shard) in a [cluster](overview.md#cluster)
+  2. **Sequence**: an ever-increasing counter
+
+!!! example "Example: Execution IDs"
+    When used in _stand-alone_ mode, the Shard ID is set to zeros ('0') so the first execution ID would be:
+
+    ```
+    0000000000000000000000000000000000000000-1
+    ```
+
+    Whereas in _cluster_ mode it could be:
+
+    ```
+    a007297351b007297351c007297351d007297351-1
+    ```
+
+## Registration
+The representation of an event-driven function is called a registration.
+
+Registrations are persisted in Redis' snapshots, a.k.a RDB files. This allows recovering both data and event handlers in the database in the event of failure.
+
+!!! abstract "Related commands"
+    The following RedisGears commands are related to registering functions:
+
+    * [`RG.PYEXECUTE`](commands.md#rgpyexecute)
+    * [`RG.DUMPREGISTRATIONS`](commands.md#rgdumpregistrations)
+    * [`RG.UNREGISTER`](commands.md#rgunregister)
+
+## Registration ID
+Every registration has a unique internal identifier. It is generated in the same manner as the [Execution ID](#execution-id).
```

Comment:
It overlap the execution ID, i.e, you might see the same id for registration and execution (again not sure if it should be mentioned ...)

Reply:
Fixed via 079f280